tests/suites/test_suite_pk.function

```diff
@@ -2375,63 +2357,118 @@ void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
     }
 
     /* Generate a 2nd PK contex using only the public key derived from its private
-     * counterpart generated above.  */
+     * counterpart generated above. */
     pub_key_id = pk_psa_pub_key_from_priv(priv_key_id, pub_key_type, key_usage, key_alg, key_bits);
     TEST_EQUAL(mbedtls_pk_copy_from_psa(pub_key_id, &pk_pub), 0);
 
-    /* Check that the 2 generated PK contexts form a valid private/public key pair. */
+    /* Test #1: check that the 2 generated PK contexts form a valid private/public key pair. */
     TEST_EQUAL(mbedtls_pk_check_pair(&pk_pub, &pk_priv, mbedtls_test_rnd_std_rand, NULL), 0);
 
-    /* Test sign/verify with the following pattern:
+    /* Test #2: sign/verify with the following pattern:
      * - Sign using the PK context generated from the private key.
      * - Verify from the same PK context used for signature.
      * - Verify with the PK context generated using public key.
+     * - Verify using the public PSA key directly.
+     *
+     * Note: mbedtls_pk_verify_ext() is tested only for RSA keys with PKCS1 v2.1
+     *      padding. For other key types this functions works the same as
+     *      mbedtls_pk_verify().
+     *
+     * Note: PSS requires the hash to be specified on sign operation (i.e. not
+     *       null or any), so in case PSA_ALG_ANY_HASH is provided as input, we
+     *       use mbedtls_pk_sign_ext() instead of mbedtls_pk_sign().
      */
-    if ((PSA_ALG_IS_RSA_OAEP(key_alg) || PSA_ALG_IS_RSA_PSS(key_alg))) {
-        mbedtls_pk_rsassa_pss_options pss_opt = {
-            .mgf1_hash_id = md_for_test,
-            .expected_salt_len = MBEDTLS_RSA_SALT_LEN_ANY,
-        };
+    mbedtls_pk_rsassa_pss_options pss_opt = {
+        .mgf1_hash_id = md_for_test,
+        .expected_salt_len = MBEDTLS_RSA_SALT_LEN_ANY,
+    };
 
+    if ((PSA_ALG_IS_RSA_OAEP(key_alg) || PSA_ALG_IS_RSA_PSS(key_alg)) &&
+        (!is_psa_hash_alg_specified)) {
         TEST_EQUAL(mbedtls_pk_sign_ext(MBEDTLS_PK_RSASSA_PSS, &pk_priv, md_for_test,
                                        in_buf, in_buf_len,
                                        out_buf, sizeof(out_buf), &out_buf_len,
                                        mbedtls_test_rnd_std_rand, NULL), 0);
+    } else {
+        TEST_EQUAL(mbedtls_pk_sign(&pk_priv, md_for_test, in_buf, in_buf_len,
```

Comment:
After Gilles' last review, `pk_sign_ext()` is basically only used in case of PSS + ANY_HASH, because we need to pass an MD alg for the signature. Should I remove this case as well as I did with OAEP+ANY_HASH?

Reply:
Mmmm I think I found another disparity between `!USE_PSA` and `USE_PSA` cases. Tests with `PSS_ANY_SALT + ANY_HASH)` and `PSS + ANY_HASH` have:
- `pk_sign()` failing when `!USE_PSA`
- `pk_sign()` succesful when `USE_PSA`

This doesn't look correct to me. Wdyt?

Reply:
Digging a bit more on this I found that the call sequence in the `!USE_PSA` case is:
`rsa_sign_wrap->mbedtls_rsa_pkcs1_sign->mbedtls_rsa_rsassa_pss_sign->rsa_rsassa_pss_sign`
and `rsa_rsassa_pss_sign` performs the check on RSA context `hash_id` to verify that it is not `MBEDTLS_MD_NONE`. Skipping this check, i.e. calling directly `rsa_rsassa_pss_sign_no_mode_check` makes `pk_sign()` pass just like the `USE_PSA` case.

Reply:
`PSS(ANY_HASH)` is just a wildcard policy, it isn't a valid algorithm. So with the key resulting from `copy_from_psa`, the hash can only come from the `md_alg` parameter of `mbedtls_pk_{sign,verify}``, which can't be `MBEDTLS_MD_NONE`. Under the hood, that means `set_padding_mode(PKCS_V21, MD_NONE).` on the RSA context.

Reply:
Ok, thanks for the explanation. I thought that `PSS(ANY_HASH)` could be valid with the following explanation:
- at PK build time the user sets `ANY_HASH` (which unfortunately translates to `MD_NONE` which does not sound the same) because he/she is not interested in setting any hash since the beginning;
- then he/she specifies the MD alg when doing the signature.

However I guess that `pk_sign_ext()` is the one that must be used for the above purpose. Isn't it?
So in this case it's the `USE_PSA` that is wrong, correct?

Reply:
`MD_NONE` means “no hash specified”, which has different meanings in different cases:

* In an RSA context, it means “I'm not saying, you'll have to pass a proper hash algorithm when you call an operation that needs one”.
* For pk sign/verify for an RSA key whose padding mode is set to PKCS#1v1.5, it means “`hash` contains a DigestInfo, not a hash”.
* For pk sign/verify for an RSA key whose padding mode is set to PKCS#1v2.1 (so PSS), it means “`use the  hash algorithm in the context (leading to an error if that is also `MD_NONE`).

``PSA_ALG_ANY_HASH` in the policy means “no hash specified” when creating the key, so it corresponds to `MBEDTLS_MD_NONE` in the RSA context.

The non-ext pk sign/verify function have a `md_alg` parameter so you don't need to use the ext functions to specify a hash.

> So in this case it's the USE_PSA that is wrong, correct?

I guess? The choice of hash algorithm has to come from somewhere. (For PSS: for v1.5 `MD_NONE` is valid as discussed above.) If there's no algorithm in the key policy and none passed to `mbedtls_pk_sign`, where does the `USE_PSA_CRYPTO` code get one?

Reply:
If `pk_sign` and friends receive a hash algorithm that isn't NONE, it always overrides the value in the context. I thought we were talking about the case of `pk_sign(..., MBEDTLS_MD_NONE, ...)`.

Reply:
Sorry, it's my fault that I didn't explain well what the issue was about. However I've just pushed a [commit](https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/b2fca8df1d2076087be6669bfc346c843533d8a0) which should make the CI happy.